gusto/initialisation_tools.py

```diff
@@ -279,3 +287,153 @@ def calculate_Pi0(state, theta0, rho0):
+              'pc_type': 'lu',
+              'pc_factor_mat_solver_package': 'mumps'}
+
+    theta0.interpolate(theta_e)
+    water_v0.interpolate(water_t)
+    Pi = Function(Vr)
+    epsilon = 0.9  # relaxation constant
+
+    # set up mixed space
+    Z = MixedFunctionSpace((Vt, Vt))
+    z = Function(Z)
+
+    gamma, phi = TestFunctions(Z)
+
+    theta_v, w_v = z.split()
+
+    # give first guesses for trial functions
+    theta_v.assign(theta0)
+    w_v.assign(water_v0)
+
+    theta_v, w_v = split(z)
+
+    # define variables
+    T = Pi * theta_v / (1 + w_v * R_v / R_d)
+    p = p_0 * Pi ** (cp / R_d)
+    L_v = L_v0 - (c_pl - c_pv) * (T - T_0)
+    w_sat = w_sat1 / (p * exp(w_sat2 * (T - T_0) / (T - w_sat3)) - w_sat4)
+
+    dxp = dx(degree=(quadrature_degree))
+
+    # set up weak form of theta_e and w_sat equations
+    F = (-gamma * theta_e * dxp
+         + gamma * T * (p / (p_0 * (1 + w_v * R_v / R_d))) **
+         (- R_d / (cp + c_pl * water_t)) *
+         exp(L_v * w_v / ((cp + c_pl * water_t) * T)) * dxp
+         - phi * w_v * dxp
+         + phi * w_sat * dxp)
+
+    problem = NonlinearVariationalProblem(F, z)
+    solver = NonlinearVariationalSolver(problem, solver_parameters=params)
+
+    theta_v, w_v = z.split()
+
+    Pi_h = Function(Vr).interpolate((p / p_0) ** (R_d / cp))
+
+    # solve for Pi with theta_v and w_v constant
+    # then solve for theta_v and w_v with Pi constant
+    for i in range(5):
+        compressible_hydrostatic_balance(state, theta0, rho0, pi0=Pi_h, water_t=water_t)
+        Pi.assign(Pi * (1 - epsilon) + epsilon * Pi_h)
+        solver.solve()
+        theta0.assign(theta0 * (1 - epsilon) + epsilon * theta_v)
+        water_v0.assign(water_v0 * (1 - epsilon) + epsilon * w_v)
+
+    # now begin on Newton solver, setup up new mixed space
+    Z = MixedFunctionSpace((Vt, Vt, Vr, Vv))
+    z = Function(Z)
+
+    gamma, phi, psi, w = TestFunctions(Z)
+
+    theta_v, w_v, pi, v = z.split()
+
+    # use previous values as first guesses for newton solver
+    theta_v.assign(theta0)
+    w_v.assign(water_v0)
+    pi.assign(Pi)
+
+    theta_v, w_v, pi, v = split(z)
+
+    # define variables
+    T = pi * theta_v / (1 + w_v * R_v / R_d)
+    p = p_0 * pi ** (cp / R_d)
+    L_v = L_v0 - (c_pl - c_pv) * (T - T_0)
+    w_sat = w_sat1 / (p * exp(w_sat2 * (T - T_0) / (T - w_sat3)) - w_sat4)
+
+    F = (-gamma * theta_e * dxp
+         + gamma * T * (p / (p_0 * (1 + w_v * R_v / R_d))) **
+         (- R_d / (cp + c_pl * water_t)) *
+         exp(L_v * w_v / ((cp + c_pl * water_t) * T)) * dxp
+         - phi * w_v * dxp
+         + phi * w_sat * dxp
+         + cp * inner(v, w) * dxp
+         - cp * div(w * theta_v / (1.0 + water_t)) * pi * dxp
+         + psi * div(theta_v * v / (1.0 + water_t)) * dxp
+         + cp * inner(w, n) * pi_boundary * theta_v / (1.0 + water_t) * ds_b
+         + g * inner(w, state.k) * dxp)
+
+    bcs = [DirichletBC(Z.sub(3), Expression(("0.", "0.")), "top")]
```

Comment:
See PR https://github.com/firedrakeproject/gusto/pull/103 for how to avoid using Expression